homework_3/src/main/java/com/epam/spring/controller/ProductController.java

```diff
@@ -0,0 +1,63 @@
+package com.epam.spring.controller;
+
+import com.epam.spring.controller.dto.ProductDto;
+import com.epam.spring.service.ProductService;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/product")
+@RequiredArgsConstructor
+@Slf4j
+public class ProductController {
+
+    private final ProductService productService;
+
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping
+    public List<ProductDto> getAllProducts(@RequestParam(required = false) String categoryName,
+                                           @RequestParam(required = false) String language) {
+        log.info("getAllProducts by categoryName (optional) {}, language (optional) {}", categoryName, language);
+        return productService.listProducts(categoryName, language);
+    }
+
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping(value = "/name/{name}")
+    public ProductDto getProductByName(@PathVariable String name, @RequestParam(required = false) String language) {
+        log.info("getProduct by name {}, language (optional) {}", name, language);
+        return productService.getProduct(name, language);
+    }
+
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping(value = "/{id}")
+    public ProductDto getProductById(@PathVariable int id, @RequestParam(required = false) String language) {
+        log.info("getProduct by id {}, language (optional) {}", id, language);
+        return productService.getProduct(id, language);
+    }
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    public ProductDto createProduct(@RequestBody ProductDto productDto) {
+        log.info("createProduct with id {}", productDto.getId());
```

Comment:
Hi Viktor. **id** retrieved from the ui is ignored since later it is generated on the Repository level and rewritten. We can just remove the id from the error message or replace it by some other information. Please do the same action on the **Service** layer. Thank you!